components/scream/src/physics/p3/tests/p3_ice_collection_unit_tests.cpp

```diff
@@ -0,0 +1,315 @@
+      for (Int s = 0; s < Spack::n; ++s) {
+        rho[s]         = rain_device(s).rho;
+        temp[s]        = rain_device(s).temp;
+        rhofaci[s]     = rain_device(s).rhofaci;
+        logn0r[s]      = rain_device(s).logn0r;
+        f1pr07[s]      = rain_device(s).f1pr07;
+        f1pr08[s]      = rain_device(s).f1pr08;
+        qitot_incld[s] = rain_device(s).qitot_incld;
+        nitot_incld[s] = rain_device(s).nitot_incld;
+        qr_incld[s]    = rain_device(s).qr_incld;
+      }
+
+      Spack qrcol(0.0), nrcol(0.0);
+      Functions::ice_rain_collection(rho, temp, rhofaci, logn0r, f1pr07, f1pr08,  
+                                     qitot_incld, nitot_incld, qr_incld,
+                                     qrcol, nrcol);
+
+
+      // Copy results back into views
+      for (Int s = 0; s < Spack::n; ++s) {
+        rain_device(s).qrcol = qrcol[s];
+        rain_device(s).nrcol = nrcol[s];
+      }
+    });
+
+    // Sync back to host
+    Kokkos::deep_copy(rain_host, rain_device);
+
+    // Validate results
+    for (Int s = 0; s < Spack::n; ++s) {
+      REQUIRE(rain[s].qrcol == rain_host(s).qrcol);
+      REQUIRE(rain[s].nrcol == rain_host(s).nrcol);
+    }
+  }
+
+  static void run_ice_rain_phys()
+  {
+    // TODO
+  }
+
+  static void run_ice_self_bfb()
+  {
+    using KTH = KokkosTypes<HostDevice>;
+
+    static constexpr Int max_pack_size = 16;
+    REQUIRE(Spack::n <= max_pack_size);
+
+    IceSelfCollectionData self[max_pack_size] = {
+     //   rho      rhofaci    f1pr03     eii       qirim      qitot        nitot
+      {4.056E-03, 8.852E-01, 0.174E+00, 1.221E-14, 5.100E-03, 9.558E-04, 9.952E+03},
+      {6.852E-02, 8.852E-01, 0.374E+00, 1.221E-13, 4.100E-03, 9.558E-04, 9.952E+03},
+      {8.852E-02, 8.900E-01, 0.123E+00, 1.221E-12, 3.100E-03, 9.558E-04, 9.952E+03},
+      {1.902E-01, 9.900E-01, 0.123E+00, 1.221E-11, 2.100E-03, 9.558E-04, 9.952E+03},
+
+      {2.201E-01, 0.100E+01, 0.174E+00, 1.221E-10, 1.100E-03, 2.558E-05, 9.952E+04},
+      {3.502E-01, 0.100E+01, 0.374E+00, 1.221E-09, 8.100E-04, 2.558E-05, 9.952E+04},
+      {4.852E-01, 0.100E+01, 0.123E+00, 1.221E-08, 4.100E-04, 2.558E-05, 9.952E+04},
+      {5.852E-01, 0.100E+01, 0.123E+00, 1.221E-07, 2.100E-04, 2.558E-05, 9.952E+04},
+
+      {6.852E-01, 0.950E+00, 0.150E+00, 1.221E-06, 9.952E-05, 4.596E-05, 1.734E+04},
+      {7.852E-01, 0.950E+00, 0.374E+00, 1.221E-05, 4.952E-05, 4.596E-05, 1.734E+04},
+      {8.852E-01, 0.950E+00, 0.123E+00, 1.221E-04, 1.952E-05, 4.596E-05, 1.734E+04},
+      {9.852E-01, 0.950E+00, 0.123E+00, 1.221E-03, 9.952E-06, 4.596E-05, 1.734E+04},
+
+      {1.002E+01, 1.069E+00, 0.174E+00, 1.221E-02, 6.952E-06, 6.596E-05, 1.734E+04},
+      {1.152E+01, 1.069E+00, 0.374E+00, 1.221E-02, 3.952E-06, 6.596E-05, 1.734E+04},
+      {1.252E+01, 1.069E+00, 0.123E+00, 1.221E-02, 1.952E-06, 6.596E-05, 1.734E+04},
+      {1.352E+01, 1.069E+00, 0.123E+00, 1.221E-02, 9.952E-07, 6.596E-05, 1.734E+04}     
+    };
+    // Get data from fortran
```

Comment:
Same comment as above.

Reply:
if the micro_p3.F90 is implemented with cxx option, do we compare the results with itself, instead of the original fortran code? This is the concern about when to add the cxx code call in micro_p3.F90.

Reply:
We've followed this pattern for all prior conversions, so it should work fine. You just need to make use_cxx is set to false when you want to call fortran and you already did that in your _c functions.